### PR TITLE
Copying dotcom attribution client to gateway

### PR DESCRIPTION
**Pull request by cbart**
Part of #59262

@keegancsmith implemented this dotcom client in the enterprise instance. We're routing search request it to gateway (will cleanup after we ship this) as per [RFC 826](https://docs.google.com/document/d/1SGzx4e-lGbXJ5hyLQz_mhiLdkhmgFuSF8uQjeRFFzQw/edit#heading=h.trqab8y0kufp).

Please also see [RFC 862](https://docs.google.com/document/d/1zSxFDQPxZcn5b6yKx40etpJayoibVzj_Gnugzln1weI/view) for more holistic albeit incomplete view of the project for February release.

## Test plan

Extend handler test to use codegen graphQL client.

### CODE REVIEW

File: cmd/cody-gateway/internal/httpapi/attribution/handler.go [hunk 1]
```diff
@@ -58,14 +60,22 @@ func NewHandler(client graphql.Client, baseLogger log.Logger) http.Handler {
 			response.JSONError(logger, w, http.StatusBadRequest, err)
 			return
 		}
-		// TODO(#59244): Actually query dotcom for attribution.
-		response := &Response{
-			TotalCount: 0,
-			LimitHit:   false,
+		searchResponse, err := dotcom.SnippetAttribution(ctx, client, request.Snippet, request.Limit)
```
Comment (chwarwick) [inline cmd/cody-gateway/internal/httpapi/attribution/handler.go:63]: I think an upper bound on Limit would be worthwhile,

**Review by chwarwick — APPROVED** [review]
looks good

File: cmd/cody-gateway/internal/dotcom/operations.go [hunk 2]
```diff
@@ -1139,6 +1139,70 @@ func (v *RateLimitFields) GetLimit() genhelper.BigInt { return v.Limit }
...
+// GetSnippetAttribution returns SnippetAttributionResponse.SnippetAttribution, and is useful for accessing the field via an interface.
+func (v *SnippetAttributionResponse) GetSnippetAttribution() SnippetAttributionSnippetAttributionSnippetAttributionConnection {
+	return v.SnippetAttribution
+}
+
+// SnippetAttributionSnippetAttributionSnippetAttributionConnection includes the requested fields of the GraphQL type SnippetAttributionConnection.
+// The GraphQL type's documentation follows.
+//
+// EXPERIMENTAL: A list of snippet attributions.
+type SnippetAttributionSnippetAttributionSnippetAttributionConnection struct {
```
[the remark this answers is not in the document]
Comment (rafax) [reply cmd/cody-gateway/internal/dotcom/operations.go:1157]: Brings back Java memories (`AbstractProxyFactory` and friends)

**Review by arafatkatze — APPROVED** [review]

File: cmd/cody-gateway/internal/httpapi/attribution/handler.go (hunk 1 above)
Comment (cbart) [reply cmd/cody-gateway/internal/httpapi/attribution/handler.go]: Done. Great point!

File: cmd/cody-gateway/internal/dotcom/operations.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbart) [reply cmd/cody-gateway/internal/dotcom/operations.go:1157]: ❤️ ☕

File: cmd/cody-gateway/internal/httpapi/attribution/handler.go [hunk 3]
```diff
@@ -9,12 +9,17 @@ import (
...
 
 	"github.com/sourcegraph/sourcegraph/cmd/cody-gateway/internal/actor"
+	"github.com/sourcegraph/sourcegraph/cmd/cody-gateway/internal/dotcom"
 	"github.com/sourcegraph/sourcegraph/cmd/cody-gateway/internal/response"
 	"github.com/sourcegraph/sourcegraph/internal/codygateway"
 	"github.com/sourcegraph/sourcegraph/internal/trace"
 	"github.com/sourcegraph/sourcegraph/lib/errors"
 )
 
+// LimitUpperBound is the maximum (and defautl) value allowed for Limit request parameter.
```
Comment (rafax) [inline cmd/cody-gateway/internal/httpapi/attribution/handler.go:19]:
```suggestion
// LimitUpperBound is the maximum (and default) value allowed for Limit request parameter.
```

File: cmd/cody-gateway/internal/httpapi/attribution/handler.go [hunk 4]
```diff
@@ -9,12 +9,17 @@ import (
...
+// If a higher value is given, then this default is set.
```
Comment (rafax) [inline cmd/cody-gateway/internal/httpapi/attribution/handler.go:20]: Any context on why you chose for (and not ex. 10)? Not implying 4 is a worse number than 10 :)

Comment (cbart) [reply cmd/cody-gateway/internal/httpapi/attribution/handler.go:20]: I just did not see a reason to request more than 4, and wanted to pick the smallest number I can think of.

**Pull request merged**